frame/recovery/src/lib.rs

```diff
@@ -0,0 +1,665 @@
+			double_map hasher(twox_64_concat) T::AccountId, hasher(twox_64_concat) T::AccountId =>
+			Option<ActiveRecovery<T::BlockNumber, BalanceOf<T>, T::AccountId>>;
+		/// The final list of recovered accounts.
+		///
+		/// Map from the recovered account to the user who can access it.
+		pub Recovered get(fn recovered_account): map T::AccountId => Option<T::AccountId>;
+	}
+}
+
+decl_event! {
+	/// Events type.
+	pub enum Event<T> where
+		AccountId = <T as system::Trait>::AccountId,
+	{
+		/// A recovery process has been set up for an account
+		RecoveryCreated(AccountId),
+		/// A recovery process has been initiated for account_1 by account_2
+		RecoveryInitiated(AccountId, AccountId),
+		/// A recovery process for account_1 by account_2 has been vouched for by account_3
+		RecoveryVouched(AccountId, AccountId, AccountId),
+		/// A recovery process for account_1 by account_2 has been closed
+		RecoveryClosed(AccountId, AccountId),
+		/// Account_1 has been successfully recovered by account_2
+		AccountRecovered(AccountId, AccountId),
+		/// A recovery process has been removed for an account
+		RecoveryRemoved(AccountId),
+	}
+}
+
+decl_error! {
+	pub enum Error for Module<T: Trait> {
+		/// User is not allowed to make a call on behalf of this account
+		NotAllowed,
+		/// Threshold must be greater than zero
+		ZeroThreshold,
+		/// Friends list must be greater than zero and threshold
+		NotEnoughFriends,
+		/// Friends list must be less than max friends
+		MaxFriends,
+		/// Friends list must be sorted and free of duplicates
+		NotSorted,
+		/// This account is not set up for recovery
+		NotRecoverable,
+		/// This account is already set up for recovery
+		AlreadyRecoverable,
+		/// A recovery process has already started for this account
+		AlreadyStarted,
+		/// A recovery process has not started for this rescuer
+		NotStarted,
+		/// This account is not a friend who can vouch
+		NotFriend,
+		/// The friend must wait until the delay period to vouch for this recovery
+		DelayPeriod,
+		/// This user has already vouched for this recovery
+		AlreadyVouched,
+		/// The threshold for recovering this account has not been met
+		Threshold,
+		/// There are still active recovery attempts that need to be closed
+		StillActive,
+		/// There was an overflow in a calculation
+		Overflow,
+	}
+}
+
+decl_module! {
+	pub struct Module<T: Trait> for enum Call where origin: T::Origin {
+		/// Deposit one of this module's events by using the default implementation.
+		fn deposit_event() = default;
+		
+		/// Send a call through a recovered account.
+		///
+		/// The dispatch origin for this call must be _Signed_ and registered to
+		/// be able to make calls on behalf of the recovered account.
+		///
+		/// Parameters:
+		/// - `account`: The recovered account you want to make a call on-behalf-of.
+		/// - `call`: The call you want to make with the recovered account.
+		///
+		/// # <weight>
+		/// - The weight of the `call`.
```

Comment:
There's a storage lookup here, too. (`Self::recovered_account`)

Reply:
Ah sorry about that. I will do another pass at weights.